### ISSUE TITLE
Strange FGSM Results with CIFAR10

### ISSUE THREAD

**Issue by inkawhich** [bethgelab/foolbox#207]
I am testing the foolbox.FGSM against my FGSM implementation in PyTorch to verify behavior and I am getting some strange results. I am only testing with one epsilon at a time and for most epsilon values, the attacks are the same and the perturbed images are equal. However, for some epsilons our results differ but the gradients match. For example, at eps=0.03 our adv images match and perturbations are barely visible, at eps=0.04 our adv images do not match but our gradients match, at eps=0.085 our results match. The most interesting part is that in the foolbox adversarial examples that do not match my version, the perturbation seems to disappear whereas my perturbations are visible as I expect.

My images are on the left, foolbox images on the right.

![image](https://user-images.githubusercontent.com/35571418/45910641-cf1a6800-bdd8-11e8-8b7e-9670924f803a.png)

![image](https://user-images.githubusercontent.com/35571418/45910643-d5a8df80-bdd8-11e8-8a4c-0cf9828c73ec.png)

![image](https://user-images.githubusercontent.com/35571418/45910645-d93c6680-bdd8-11e8-899c-b021e885d066.png)

Do you have any idea what may be happening here? Is this intended behavior and I am just missing something?

Thanks,
Nate

**Comment by wielandbrendel** [comment]
That's indeed a bit strange. What Foolbox does internally is to save the smallest adversarial it encountered throughout an attack. If no adversarial is found it returns None. Could you post the code snippet that shows how you call the Foolbox attack and get the result? Also, it might be worth for you to quickly copy the attack into your script (the FGSM is a very simple attack) and then go through the mechanics yourself (i.e. check exactly what gradients are generated, where and how they are added, and what the model returns on the perturbed image.

**Comment by inkawhich** [comment]
Wieland,

This is how I call the foolbox attack, where image is a CHW np.array range [0,1]. I believe this calls a single step of the attack with strength eps. 

```
mean = np.array([0.4914, 0.4822, 0.4465]).reshape((3, 1, 1)) 
std = np.array([0.2023, 0.1994, 0.2010]).reshape((3, 1, 1)) 
fmodel = foolbox.models.PyTorchModel(whitebox_model, bounds=(0, 1), num_classes=10, preprocessing=(mean, std))
print('predicted class', np.argmax(fmodel.predictions(image)))
fbgrad = fmodel.gradient(image,label)
attack = foolbox.attacks.FGSM(fmodel)
adversarial = attack(image, label, epsilons=[eps])
print('adversarial class', np.argmax(fmodel.predictions(adversarial)))
```

For reference, here is my FGSM implementation:

```
def fgsm_attack(model, device, data, lbl, eps):

    # Cifar-10 normalization
    mean = torch.tensor([0.4914, 0.4822, 0.4465], dtype=torch.float32)
    std = torch.tensor([0.2023, 0.1994, 0.2010], dtype=torch.float32)
    normalize = transforms.Normalize(mean.tolist(), std.tolist())

    #### ATOMIC OPERATION: (Normalize data, forward pass, backward pass, collect gradient)  
    # Format data and label as tensors and send to device
    dat = normalize(torch.from_numpy(np.copy(data))).unsqueeze_(0).to(device)
    cls = torch.from_numpy(np.array([lbl])).to(device)
    # Forward pass through the model
    dat.requires_grad = True
    out = model(dat)
    # Calculate loss
    loss = F.cross_entropy(out,cls)
    # zero all old gradients in the model
    model.zero_grad()
    # Back prop the loss to calculate gradients
    loss.backward()
    # Extract gradient of loss w.r.t data
    data_grad = dat.grad.data.cpu().numpy().squeeze()
    # Unnorm gradients back into [0,1] space
    grad = data_grad / std.numpy().reshape((3,1,1)) 
    #### END ATOMIC OPERATION

    # Collect the element-wise sign of the data gradient
    sign_data_grad = np.sign(grad)
    # Create the perturbed image by adjusting each pixel of the input image
    perturbed_image = data + eps*sign_data_grad
    # Adding clipping to maintain original data range
    perturbed_image = np.clip(perturbed_image, 0., 1.) 

    # Return a numpy array of the perturbed input data
    return perturbed_image, grad
```

**Comment by jonasrauber** [comment]
@inkawhich Could you provide a minimal example that's fully executable (maybe a Jupyter notebook), that shows how you run Foolbox on a particular model and it provides something else than you would expect?

**Comment by inkawhich** [comment]
@jonasrauber I will try to do this when I get a chance. I noticed that it is also image dependent, so different images give different results. For this reason it may be hard to recreate unless you have my models and stuff. I will try to figure something out soon. Just to make sure, do you agree with my FGSM version shown above?

**Comment by jonasrauber** [comment]
Closing due to lack of activity. Please reopen if the issue still exits.
